### PR TITLE
Avoid async infinite loop on tab close

### PR DESCRIPTION
**Pull request by lhecker**
When a tab gets closed, `_RemoveTab` will call `TabBase::Shutdown()`,
which then re-raises the `Closed` event, which will end up calling
`_RemoveTab` again, etc. The only reason this didn't crash WT so far
is because `_RemoveOnCloseRoutine` contains a `resume_foreground`,
which would resolve the recursion and turn it into CPU usage.
It would spin as long as WinUI hasn't discard the tab object,
which takes an unpredictable amount of time.

Raising the `Closed` event from `Shutdown()` is unnecessary, because
the handlers of the event end up calling `_RemoveTab` anyways.
Technically the entire `Closed` event can be removed now, but I left it
in anyways because resolving the architectural "knot" around the way
tab closing after the last pane closes is implemented requires much
more significant changes.

This commit additionally removes the `_createCloseLock` mutex in `Pane`
as it was very likely not working as intended anyways. Only some methods
were protected by it and it doesn't avoid any STA/MTA/NA issues either.

## Validation Steps Performed
* Closing tabs and panes always ends up calling `Shutdown()` ✅

### CODE REVIEW

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 1]
```diff
@@ -509,6 +512,11 @@ namespace winrt::TerminalApp::implementation
             _mruTabs.RemoveAt(mruIndex);
         }
 
+        if (tab == _settingsTab)
+        {
+            _settingsTab = nullptr;
+        }
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TabManagement.cpp:518]: I felt like this is a more robust way to express the same idea of "there's just one instance of settings and closing it destroys it".

**Review by DHowett — CHANGES_REQUESTED** [review]
What about when a tab _closes itself_, like when the terminal connection inside it exits? That does not come in on the main dispatcher thread, therefore interacting with UI elements during the Close callback is a danger.

Are we certain that Close always comes in on the _right_ thread?

File: src/cascadia/TerminalApp/Pane.cpp [hunk 2]
```diff
@@ -1843,135 +1842,128 @@ void Pane::_CloseChild(const bool closeFirst, const bool isDetaching)
     closedChild->_ClosedByParentHandlers();
 }
 
-winrt::fire_and_forget Pane::_CloseChildRoutine(const bool closeFirst)
+void Pane::_CloseChildRoutine(const bool closeFirst)
```
Comment (lhecker) [inline src/cascadia/TerminalApp/Pane.cpp:1845]: You need to view this changeset with whitespace changes suppressed. :)

**Comment by lhecker** [comment]
> Are we certain that Close always comes in on the right thread?

You're right, it didn't. In my original PR it worked, because I also made it so that the close event gets emitted on the UI thread. I now brought those changes into this PR as well, but unfortunately it increases the diff size. It's easier to view with whitespace changes suppressed.

**Review by DHowett — APPROVED** [review]
I think I understand why this works. Thanks!

1.18 material?

**Comment by DHowett** [comment]
Oh, you _removed_ it from 1.18. OK

**Comment by lhecker** [comment]
Yeah I think I'd like to wait with this until 1.19... There's still these comments in the code and I'd like to 110% grasp what they mean in practice and if it was a practical concern to begin with:
```
// It's possible that this event handler started being executed, then before
// we got the lock, another thread created another child. So our control is
// actually no longer _our_ control, and instead could be a descendant.
```
...because to be honest I don't think any of this worked the way we imagined it would.

**Review by carlos-zamora — COMMENTED** [review]
This all seems good. The only thing I want to double check with you on is why we're removing the lock from...
- `SwapPanes()`
- `_CloseTerminalRequestedHandler()`
- `Shutdown()`
- `_CloseChild()`
- `_CloseChildRoutine()`

I get that's the point of this PR haha, but why is it safe to do that now?

**Comment by lhecker** [comment]
@carlos-zamora Sorry for the late response. As far as I can tell, `_ControlConnectionStateChangedHandler` is the only function that is being called from a background thread. It would end up calling `Close()` which I assume is why the mutex was called `_createCloseLock`.

The reason I removed it is simply because that mutex provides a false sense of security. We can't use it to make shutdown secure, when Pane has ~50 member functions and only 6 of them are thread safe. 😅 It just doesn't work at all and never did. 😅😅

**Review by carlos-zamora — APPROVED** [review]

**Comment by zadjii-msft** [comment]
This has ✅✅, but never merged while I was out. I haven't read it yet. Should we include this in the bug bash?

**Comment by lhecker** [comment]
Yes, please. I've merged in main just now to resolve the conflict.

**Pull request merged**